service/service.go

```diff
@@ -412,21 +412,29 @@ func (s *Service) executeRound(ctx context.Context, r *round) error {
 // Temporarily support both the old and new challenge submission API.
 // TODO(brozansk) remove support for data []byte after go-spacemesh is updated to
 // use the new API.
+// (https://github.com/spacemeshos/poet/issues/147).
 func (s *Service) Submit(ctx context.Context, challenge []byte, signedChallenge signing.Signed[shared.Challenge]) (*round, []byte, error) {
 	if !s.Started() {
 		return nil, nil, ErrNotStarted
 	}
 
-	// TODO(brozansk) Remove support for `challenge []byte` eventually
+	// TODO(brozansk) Remove support for `challenge []byte` eventually (https://github.com/spacemeshos/poet/issues/147)
 	if signedChallenge != nil {
 		log.Debug("Using the new challenge submission API")
 		// SAFETY: it will never panic as `s.atxProvider` is set in Start
 		atxProvider := s.atxProvider.Load().(types.AtxProvider)
 		if err := validateChallenge(ctx, signedChallenge, atxProvider, s.postConfig.Load(), verifying.Verify); err != nil {
```

Comment:
since `atxProvider` and `postConfig` can be changed independently; one or both might change between `atxProvider.Load()` and `postConfig.Load()`.

I think these two (along possibly with `boradcaster` should be changed atomically or rare issues might occur where the wrong postConfig is used to validate a challenge from a different atxProvider.

Reply:
`atxProvider` doesn't provide challenges, but ATXes for a submitted challenge. It's perfectly fine if a `postConfig` comes from a different Gateway than the ATXes fetched from an atxProvider. Anyway, I plan to fix all data races when I rewrite Poet Service to an actor.